network/gossip/libp2p/topology/randPermTopology.go

```diff
@@ -0,0 +1,66 @@
+package topology
+
+import (
+	"fmt"
+
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/model/flow/filter"
+)
+
+var _ Topology = &RandPermTopology{}
+
+// RandPermTopology generates a random topology from a given set of nodes and for a given role
+// The topology generated is a union of three sets:
+// 1. a random subset of the size (n+1)/2 to make sure the nodes form a connected graph with no islands
+// 2. one node of each of the flow role from the remaining ids to make a node can talk to any other type of node
+// 3. (n+1)/2 of the nodes of the same role as this node from the remaining ids to make sure that nodes of the same type
+// form a connected graph with no islands.
+type RandPermTopology struct {
+	myRole flow.Role
+	seed   int64
+}
+
+func NewRandPermTopology(role flow.Role, id flow.Identifier) (RandPermTopology, error) {
+	seed, err := seedFromID(id)
+	if err != nil {
+		return RandPermTopology{}, fmt.Errorf("failed to seed topology: %w", err)
```

Comment:
The `RandPermTopology` is immutable such that the role and the id are not suppose to be changed, hence the return by value instead of reference (in addition to the role and id being private).